### PR TITLE
[EDGINREACH-16] EDGE: Implement a new error response type

### PR DESCRIPTION
**Pull request by aleksandr2110**
Purpose
https://issues.folio.org/browse/EDGINREACH-16
The documentations describes two types of error responces:

The first one implemented now is for authentication errors. (D2IR API Reference v2.4, page 9)
The second one is for common errors (D2IR API Reference v2.4, Record contrinution, Circulation chapters)
We need to implement the second option as well, as well as refactor the code to use the appropriate error response for different cases.

Hint for devs: This error response type is already implemented in mod-inn-reach
Approach

## Pre-Merge Checklist:
 Before merging this PR, please go through the following list and take appropriate actions.

 - Does this PR meet or exceed the expected quality standards?
   - [x] Code coverage on new code is 80% or greater
   - [x] Duplications on new code is 3% or less
   - [x] There are no major code smells or security issues
 - Does this introduce breaking changes?
   - [ ] Were any API paths or methods changed, added or removed?
   - [ ] Were there any schema changes?
   - [ ] Did any of the interface versions change?
   - [ ] Were permissions changed, added, or removed?
   - [ ] Are there new interface dependencies?
   - [ ] There are no breaking changes in this PR.

 If there are breaking changes, please **STOP** and consider the following:

 - What other modules will these changes impact?
 - Do JIRAs exist to update the impacted modules?
   - [ ] If not, please create them
   - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
   - [ ] Do they have all they appropriate links to blocked/related issues?
 - Are the JIRAs under active development?
   - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
 - Do PRs exist for these changes?
   - [ ] If so, have they been approved?

 Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.

 While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/edge/domain/exception/InnReachErrorMapper.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  @Mapping(target = "name", source = "error.field")
+  @Mapping(target = "rejectedValue", source = "error.rejectedValue")
+  @Mapping(target = "reason", source = "error.defaultMessage") InnReachError toInnReachError(FieldError error);
+
+  @Mapping(target = "reason", source = "e.message") InnReachError toInnReachError(Exception e);
+
+  default String map(Object value){
+    return value.toString();
+  }
+}
```
Comment (oliinyko) [inline src/main/java/org/folio/edge/domain/exception/InnReachErrorMapper.java:22]: is it used anywhere?

File: src/main/java/org/folio/edge/controller/exception/ExceptionHandlerController.java [hunk 2]
```diff
@@ -77,4 +83,17 @@ public Error handleException(Exception e) {
...
+  @ExceptionHandler(FeignException.class)
+  public ResponseEntity<InnReachResponseDTO> handleFeignStatusException(FeignException feignException) {
+    var status = HttpStatus.resolve(feignException.status());
+    String body = feignException.contentUTF8();
+    InnReachResponseDTO innReachResponseDTO = null;
+    try {
+      innReachResponseDTO = objectMapper.readValue(body, InnReachResponseDTO.class);
+    } catch (JsonProcessingException e) {
+      e.printStackTrace();
+    }
```
Comment (oliinyko) [inline src/main/java/org/folio/edge/controller/exception/ExceptionHandlerController.java:95]: The `innReachResponseDTO` in this case is not initialized, which will lead to null response. Prepare an error manually as a fall-back in catch block

File: src/main/java/org/folio/edge/controller/exception/ExceptionHandlerController.java [hunk 3]
```diff
@@ -77,4 +83,17 @@ public Error handleException(Exception e) {
...
+
+    return new ResponseEntity<InnReachResponseDTO>(innReachResponseDTO, status);
+  }
```
Comment (oliinyko) [inline src/main/java/org/folio/edge/controller/exception/ExceptionHandlerController.java:98]: add some logging
`
    log.error("Unexpected exception: " + e.getMessage(), e);
`

File: src/main/resources/swagger.api/schemas/v2/innReachError.json [hunk 4]
```diff
@@ -0,0 +1,37 @@
+{
+  "description": "V2 error",
```
Comment (oliinyko) [inline src/main/resources/swagger.api/schemas/v2/innReachError.json:2]: D2IR error

File: src/main/resources/swagger.api/schemas/v2/innReachResponseDTO.json [hunk 5]
```diff
@@ -0,0 +1,28 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "description": "V2 response",
```
Comment (oliinyko) [inline src/main/resources/swagger.api/schemas/v2/innReachResponseDTO.json:3]: D2IR response

File: src/main/resources/swagger.api/edge-inn-reach.yaml [hunk 6]
```diff
@@ -51,6 +51,10 @@ components:
       $ref: schemas/accessTokenResponse.json
     error:
       $ref: schemas/error.json
+    innReachError:
+      $ref: schemas/v2/innReachError.json
```
Comment (oliinyko) [inline src/main/resources/swagger.api/edge-inn-reach.yaml:55]: schemas/**d2ir**/

File: src/main/resources/swagger.api/edge-inn-reach.yaml [hunk 7]
```diff
@@ -51,6 +51,10 @@ components:
...
+    innReachEdgeResponseDTO:
+      $ref: schemas/v2/innReachResponseDTO.json
```
Comment (oliinyko) [inline src/main/resources/swagger.api/edge-inn-reach.yaml:57]: same as above

**Comment by dmtkachenko** [comment]
Any tests?

File: src/main/resources/swagger.api/edge-inn-reach.yaml (hunk 6 above)
Comment (aleksandr2110) [reply src/main/resources/swagger.api/edge-inn-reach.yaml]: fixed

**Comment by aleksandr2110** [comment]
> 

I wrote 1 test but it didn't cover the functionality

File: src/test/java/org/folio/edge/controller/InnReachProxyControllerTest.java [hunk 8]
```diff
@@ -98,4 +101,22 @@ void shouldUpdateRequestHeaders() {
     );
   }
 
+  @Test
+  void shouldHandleCommonErrors() {
+    var httpHeaders = createInnReachHttpHeaders();
+    var requestEntity = new HttpEntity<>(httpHeaders);
+
+    wireMock.stubFor(post(PATRON_VERIFY_URL_PATTERN)
+      .willReturn(aResponse().withBody("{}")
```
Comment (oliinyko) [inline src/test/java/org/folio/edge/controller/InnReachProxyControllerTest.java:110]: set some INN-reach error as body and assert if it serializes properly

File: src/main/java/org/folio/edge/controller/exception/ExceptionHandlerController.java [hunk 9]
```diff
@@ -77,4 +83,25 @@ public Error handleException(Exception e) {
...
+  public ResponseEntity<InnReachResponseDTO> handleFeignStatusException(FeignException feignException) {
+    var status = HttpStatus.resolve(feignException.status());
+    String body = feignException.contentUTF8();
+    InnReachResponseDTO innReachResponseDTO = null;
+    try {
+      innReachResponseDTO = objectMapper.readValue(body, InnReachResponseDTO.class);
+    } catch (JsonProcessingException e) {
+      log.error("Unexpected exception: {}" + e.getMessage());
+      innReachResponseDTO = failed(body);
+      return new ResponseEntity<>(innReachResponseDTO, status);
```
Comment (oliinyko) [inline src/main/java/org/folio/edge/controller/exception/ExceptionHandlerController.java:96]: return is redundant in this block, it's just a duplicate of method's return

File: src/main/java/org/folio/edge/controller/exception/ExceptionHandlerController.java [hunk 10]
```diff
@@ -77,4 +83,25 @@ public Error handleException(Exception e) {
     return new ResponseEntity<>(new Error().error(INVALID_REQUEST_ERROR_TYPE), HttpStatus.valueOf(e.getStatus()));
   }
 
+  @ExceptionHandler(FeignException.class)
+  public ResponseEntity<InnReachResponseDTO> handleFeignStatusException(FeignException feignException) {
+    var status = HttpStatus.resolve(feignException.status());
```
Comment (oliinyko) [inline src/main/java/org/folio/edge/controller/exception/ExceptionHandlerController.java:88]: add logging of exception similar to what we do for other handlers

File: src/main/java/org/folio/edge/controller/exception/ExceptionHandlerController.java (hunk 9 above)
Comment (aleksandr2110) [reply src/main/java/org/folio/edge/controller/exception/ExceptionHandlerController.java:96]: Fixed

File: src/main/java/org/folio/edge/controller/exception/ExceptionHandlerController.java (hunk 3 above)
Comment (aleksandr2110) [reply src/main/java/org/folio/edge/controller/exception/ExceptionHandlerController.java]: Fixed

File: src/test/java/org/folio/edge/controller/InnReachProxyControllerTest.java (hunk 8 above)
Comment (aleksandr2110) [reply src/test/java/org/folio/edge/controller/InnReachProxyControllerTest.java]: Fixed

File: src/main/java/org/folio/edge/controller/exception/ExceptionHandlerController.java (hunk 10 above)
Comment (aleksandr2110) [reply src/main/java/org/folio/edge/controller/exception/ExceptionHandlerController.java:90]: Fixed

File: src/test/java/org/folio/edge/controller/InnReachProxyControllerTest.java [hunk 11]
```diff
@@ -98,4 +107,72 @@ void shouldUpdateRequestHeaders() {
     );
   }
 
+  @Test
+  void shouldHandleCommonErrors() {
+    var httpHeaders = createInnReachHttpHeaders();
+    httpHeaders.set(ACCEPT_ENCODING, "gzip");
```
Comment (oliinyko) [inline src/test/java/org/folio/edge/controller/InnReachProxyControllerTest.java:113]: redundant header

File: src/test/java/org/folio/edge/controller/InnReachProxyControllerTest.java [hunk 12]
```diff
@@ -98,4 +107,72 @@ void shouldUpdateRequestHeaders() {
...
+    InnReachResponseDTO innReachResponseDTO = new InnReachResponseDTO();
+    innReachResponseDTO.status("failed");
+    ObjectMapper obj = new ObjectMapper();
+    String jsonStr = null;
+    try {
+      jsonStr = obj.writeValueAsString(innReachResponseDTO);
+    }
+    catch (IOException e) {
+      e.printStackTrace();
+    }
```
Comment (oliinyko) [inline src/test/java/org/folio/edge/controller/InnReachProxyControllerTest.java:125]: save a response as json file in src/test/resources/... then read it with `TestUtil.readFileContentAsString`. Add some "reason" value as well

File: src/test/java/org/folio/edge/controller/InnReachProxyControllerTest.java [hunk 13]
```diff
@@ -98,4 +107,72 @@ void shouldUpdateRequestHeaders() {
...
+    assertTrue(responseEntity.getStatusCode().is4xxClientError());
+    assertEquals(BAD_REQUEST, responseEntity.getStatusCode());
+    assertTrue(responseEntity.hasBody());
+    assertEquals(innReachResponseDTO, responseEntity.getBody());
+  }
+
+  @Test
+  void shouldHandleCommonErrorsWhenCantReadBody() {
+    var httpHeaders = createInnReachHttpHeaders();
+    httpHeaders.set(ACCEPT_ENCODING, "gzip");
```
Comment (oliinyko) [inline src/test/java/org/folio/edge/controller/InnReachProxyControllerTest.java:147]: redundant header

File: src/test/java/org/folio/edge/controller/InnReachProxyControllerTest.java [hunk 14]
```diff
@@ -98,4 +107,72 @@ void shouldUpdateRequestHeaders() {
...
+      e.printStackTrace();
+    }
+
+    HttpEntity<InnReachResponseDTO> requestEntity = new HttpEntity<>(innReachResponseDTO, httpHeaders);
+
+    wireMock.stubFor(post(PATRON_VERIFY_URL_PATTERN)
+      .willReturn(aResponse().withBody(jsonStr)
+        .withStatus(400)
+        .withHeader(CONTENT_TYPE, APPLICATION_JSON_VALUE)
+        .withHeader(X_OKAPI_TOKEN, TEST_TOKEN)));
```
Comment (oliinyko) [inline src/test/java/org/folio/edge/controller/InnReachProxyControllerTest.java:168]: As I understand, it should verify the fall-back scenario when json can't be serialized. Why do we need `innReachResponseDTO` and `innReachRequest` objects for this? Simplify it to something like:
```
wireMock.stubFor(post(PATRON_VERIFY_URL_PATTERN)
      .willReturn(aResponse().withBody("Some plain text")
        .withStatus(400)
        .withHeader(CONTENT_TYPE, TEXT_PLAIN_VALUE)
        .withHeader(X_OKAPI_TOKEN, TEST_TOKEN))); 
```

File: src/test/java/org/folio/edge/controller/InnReachProxyControllerTest.java (hunk 11 above)
Comment (aleksandr2110) [reply src/test/java/org/folio/edge/controller/InnReachProxyControllerTest.java:113]: Fixed

File: src/test/java/org/folio/edge/controller/InnReachProxyControllerTest.java (hunk 12 above)
Comment (aleksandr2110) [reply src/test/java/org/folio/edge/controller/InnReachProxyControllerTest.java:125]: Fixed

File: src/test/java/org/folio/edge/controller/InnReachProxyControllerTest.java (hunk 13 above)
Comment (aleksandr2110) [reply src/test/java/org/folio/edge/controller/InnReachProxyControllerTest.java:147]: Fixed

**Review by oliinyko — APPROVED** [review]

**Pull request merged**